pennylane_lightning/core/src/simulators/lightning_gpu/catalyst/LightningGPUSimulator.cpp

```diff
@@ -0,0 +1,557 @@
+        *dst = std::move(*src);
+        *src = std::complex<double>(.0, .0);
+    }
+
+    this->device_sv = std::make_unique<StateVectorT>(data.data(), data.size());
+    return this->qubit_manager.Allocate(num_qubits);
+}
+
+auto LightningGPUSimulator::AllocateQubits(std::size_t num_qubits)
+    -> std::vector<QubitIdType> {
+    if (!num_qubits) {
+        return {};
+    }
+
+    // at the first call when num_qubits == 0
+    if (!this->GetNumQubits()) {
+        this->device_sv = std::make_unique<StateVectorT>(num_qubits);
+        return this->qubit_manager.AllocateRange(0, num_qubits);
+    }
+
+    std::vector<QubitIdType> result(num_qubits);
+    std::generate_n(result.begin(), num_qubits,
+                    [this]() { return AllocateQubit(); });
+    return result;
+}
+
+void LightningGPUSimulator::ReleaseQubit(QubitIdType q) {
+    this->qubit_manager.Release(q);
+}
+
+void LightningGPUSimulator::ReleaseAllQubits() {
+    this->qubit_manager.ReleaseAll();
+    this->device_sv = std::make_unique<StateVectorT>(0); // reset the device
+}
+
+auto LightningGPUSimulator::GetNumQubits() const -> std::size_t {
+    return this->device_sv->getNumQubits();
+}
+
+void LightningGPUSimulator::StartTapeRecording() {
+    RT_FAIL_IF(this->tape_recording, "Cannot re-activate the cache manager");
+    this->tape_recording = true;
+    this->cache_manager.Reset();
+}
+
+void LightningGPUSimulator::StopTapeRecording() {
+    RT_FAIL_IF(!this->tape_recording,
+               "Cannot stop an already stopped cache manager");
+    this->tape_recording = false;
+}
+
+auto LightningGPUSimulator::CacheManagerInfo()
+    -> std::tuple<std::size_t, std::size_t, std::size_t,
+                  std::vector<std::string>, std::vector<ObsIdType>> {
+    return {this->cache_manager.getNumOperations(),
+            this->cache_manager.getNumObservables(),
+            this->cache_manager.getNumParams(),
+            this->cache_manager.getOperationsNames(),
+            this->cache_manager.getObservablesKeys()};
+}
+
+void LightningGPUSimulator::SetDeviceShots(std::size_t shots) {
+    this->device_shots = shots;
+}
+
+auto LightningGPUSimulator::GetDeviceShots() const -> std::size_t {
+    return this->device_shots;
+}
+
+void LightningGPUSimulator::SetDevicePRNG(std::mt19937 *gen) {
+    this->gen = gen;
+}
+
+/// LCOV_EXCL_START
+void LightningGPUSimulator::PrintState() {
+    using std::cout;
+    using std::endl;
+
+    const std::size_t num_qubits = this->device_sv->getNumQubits();
+    const std::size_t size = Pennylane::Util::exp2(num_qubits);
+
+    std::vector<std::complex<double>> state(size, {0.0, 0.0});
+
+    this->device_sv->CopyGpuDataToHost(state.data(), size);
+
+    std::size_t idx = 0;
+    cout << "*** State-Vector of Size " << size << " ***" << endl;
+    cout << "[";
+    for (; idx < size - 1; idx++) {
+        cout << state[idx] << ", ";
+    }
+    cout << state[idx] << "]" << endl;
+}
+/// LCOV_EXCL_STOP
```

Comment:
Is this something we would like to keep? Any special reason for that? Why skip coverage?

Reply:
It's feature parity with LK.

Reply:
Why are we skipping coverage? :thinking:

Reply:
Hey @AmintorDusko, good one. As discussed with @maliasadi, we will come back to this for all sv backends later. Happy to add TODO for this.